File: java/src/test/java/org/libpetri/export/CrossLangInhibitorReadDot.java
package org.libpetri.export;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

import org.junit.jupiter.api.Test;
import org.libpetri.core.Arc.In;
import org.libpetri.core.Arc.Out;
import org.libpetri.core.PetriNet;
import org.libpetri.core.Place;
import org.libpetri.core.Transition;

/**
 * Cross-language DOT byte-parity fixture: builds a net with at least one
 * Read arc and one Inhibitor arc on the same transition to verify the
 * renderer styles each distinctly and produces byte-identical DOT across
 * languages.
 *
 * <p>Consumed by {@code scripts/cross-lang-dot-parity.sh}.
 *
 * <p>Mirrors {@code typescript/scripts/cross-lang-inhibitor-read-dot.ts} and
 * {@code rust/libpetri/tests/cross_lang_inhibitor_read_dot.rs}.
 */
public class CrossLangInhibitorReadDot {

    /** Default output path, relative to the {@code java/} module root. */
    public static final String DEFAULT_OUT = "target/cross-lang-inhibitor-read-dot/java.dot";

    @Test
    void emitsInhibitorReadStylingDot() throws IOException {
        Place<String> req = Place.of("req", String.class);
        Place<String> done = Place.of("done", String.class);
        Place<String> lock = Place.of("lock", String.class);
        Place<String> paused = Place.of("paused", String.class);

        Transition process = Transition.builder("process")
            .inputs(In.one(req))
            .read(lock)
            .inhibitor(paused)
            .outputs(Out.place(done))
            .build();

        // Place ordering matches the renderer's place-analysis traversal
        // (inputs -> outputs -> inhibitors -> reads), required for byte-parity
        // because some language backends ignore the builder's place list and
        // re-derive ordering from arc iteration.
        var net = PetriNet.builder("inh_read")
            .place(req)
            .place(done)
            .place(paused)
            .place(lock)
            .transition(process)
            .build();

        String dot = DotExporter.export(net);

        String outPath = System.getProperty("libpetri.crossLang.outFile", DEFAULT_OUT);
        Path target = Path.of(outPath);
        if (target.getParent() != null) {
            Files.createDirectories(target.getParent());
        }
        Files.writeString(target, dot);
    }
}

File: rust/libpetri/tests/cross_lang_inhibitor_read_dot.rs
//! Cross-language DOT byte-parity fixture: builds a net with at least one
//! Read arc and one Inhibitor arc on the same transition to verify the
//! renderer styles each distinctly and produces byte-identical DOT across
//! languages.
//!
//! Mirrors `java/src/test/java/org/libpetri/export/CrossLangInhibitorReadDot.java`
//! and `typescript/scripts/cross-lang-inhibitor-read-dot.ts`. Consumed by
//! `scripts/cross-lang-dot-parity.sh`.

use std::env;
use std::fs;
use std::path::PathBuf;

use libpetri::core::arc::{inhibitor, read};
use libpetri::core::input::one;
use libpetri::core::output::out_place;
use libpetri::core::petri_net::PetriNet;
use libpetri::core::place::Place;
use libpetri::core::transition::Transition;
use libpetri_export::dot_exporter::dot_export;

fn workspace_root() -> PathBuf {
    let manifest = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    manifest
        .parent()
        .map(PathBuf::from)
        .unwrap_or(manifest)
}

#[test]
fn emits_inhibitor_read_styling_dot() {
    let req = Place::<String>::new("req");
    let done = Place::<String>::new("done");
    let lock = Place::<String>::new("lock");
    let paused = Place::<String>::new("paused");

    let process = Transition::builder("process")
        .input(one(&req))
        .read(read(&lock))
        .inhibitor(inhibitor(&paused))
        .output(out_place(&done))
        .build();

    // Place ordering matches the renderer's place-analysis traversal
    // (inputs -> outputs -> inhibitors -> reads), required for byte-parity
    // because some language backends ignore the builder's place list and
    // re-derive ordering from arc iteration.
    let net = PetriNet::builder("inh_read")
        .place(req.as_ref())
        .place(done.as_ref())
        .place(paused.as_ref())
        .place(lock.as_ref())
        .transition(process)
        .build();

    let dot = dot_export(&net, None);

    let out_path = match env::var("LIBPETRI_CROSS_LANG_OUT") {
        Ok(p) => PathBuf::from(p),
        Err(_) => workspace_root().join("target/cross-lang-inhibitor-read-dot/rust.dot"),
    };
    if let Some(parent) = out_path.parent() {
        fs::create_dir_all(parent).expect("create parent dir for cross-lang DOT output");
    }
    fs::write(&out_path, &dot).expect("write cross-lang DOT output");
}
